contracts/upgrader/src/contract.rs

```diff
@@ -0,0 +1,408 @@
+#![allow(dead_code)]
+
+use crate::error::ContractError;
+use axelar_soroban_std::traits::ThenOk;
+use soroban_sdk::{contract, contractimpl, Address, BytesN, Env, String, Symbol, Val};
+
+const MIGRATE: &str = "migrate";
+const UPGRADE: &str = "upgrade";
+const VERSION: &str = "version";
+
+#[contract]
+pub struct Upgrader;
+
+#[contractimpl]
+impl Upgrader {
+    pub fn __constructor(_env: Env) {}
+
+    pub fn upgrade(
+        env: Env,
+        caller: Address,
+        contract_address: Address,
+        new_version: String,
+        new_wasm_hash: BytesN<32>,
+        migration_data: soroban_sdk::Vec<Val>,
+    ) -> Result<(), ContractError> {
+        caller.require_auth();
+
+        // run the upgrade and migration in the context of the given caller,
+        // so to the upgradeable contract it looks like the caller invoked those functions directly,
+        // and can authenticate them correctly
+        env.as_contract(&caller, || {
+            ensure_new_version_is_different(&env, &contract_address, &new_version)?;
+
+            // it's safe to map the true return value of the upgrade and migrate functions to (),
+            // because we don't care about it, and in case of failure the contract will panic anyway
+
+            env.invoke_contract::<()>(
+                &contract_address,
+                &Symbol::new(&env, UPGRADE),
+                soroban_sdk::vec![&env, new_wasm_hash.into()],
+            );
+
+            env.invoke_contract::<()>(
```

Comment:
import the client so you can avoid using invoke_contract, e.g.
https://github.com/axelarnetwork/axelar-cgp-soroban/blob/527c2fb63b9ad1596d131fdb5baf0b0c5e7e8a95/contracts/example/src/contract.rs#L54

Reply:
as discussed, this has to be generic (not in the generic type sense), because the upgrader doesn't know which concrete error type it has to cast to